File: src/api/java/Params.java
/**
Copyright (c) 2012-2014 Microsoft Corporation
   
Module Name:

    Params.java

Abstract:

Author:

    @author Christoph Wintersteiger (cwinter) 2012-03-15

Notes:
    
**/ 


package com.microsoft.z3;

/**
 * A ParameterSet represents a configuration in the form of Symbol/value pairs.
 **/
public class Params extends Z3Object {
    /**
     * Adds a parameter setting.
     **/
    public void add(Symbol name, boolean value)
    {
        Native.paramsSetBool(getContext().nCtx(), getNativeObject(),
                name.getNativeObject(), (value));
    }

    /**
     * Adds a parameter setting.
     **/
    public void add(Symbol name, double value)
    {
        Native.paramsSetDouble(getContext().nCtx(), getNativeObject(),
                name.getNativeObject(), value);
    }
    
    /**
     * Adds a parameter setting.
     **/
    public void add(Symbol name, String value)
    {

        Native.paramsSetSymbol(getContext().nCtx(), getNativeObject(),
                name.getNativeObject(), 
                getContext().mkSymbol(value).getNativeObject());
    }

    /**
     * Adds a parameter setting.
     **/
    public void add(Symbol name, Symbol value)
    {

        Native.paramsSetSymbol(getContext().nCtx(), getNativeObject(),
                name.getNativeObject(), value.getNativeObject());
    }

    /**
     * Adds a parameter setting.
     **/
    public void add(String name, boolean value)
    {
        Native.paramsSetBool(getContext().nCtx(), getNativeObject(), 
                getContext().mkSymbol(name).getNativeObject(), value);
    }

    /**
     * Adds a parameter setting.
     **/
    public void add(String name, int value)
    {
        Native.paramsSetUint(getContext().nCtx(), getNativeObject(), getContext()
                .mkSymbol(name).getNativeObject(), value);
    }

    /**
     * Adds a parameter setting.
     **/
    public void add(String name, double value)
    {
        Native.paramsSetDouble(getContext().nCtx(), getNativeObject(), getContext()
                .mkSymbol(name).getNativeObject(), value);
    }

    /**
     * Adds a parameter setting.
     **/
    public void add(String name, Symbol value)
    {
        Native.paramsSetSymbol(getContext().nCtx(), getNativeObject(), getContext()
                .mkSymbol(name).getNativeObject(), value.getNativeObject());
    }

    /**
     * Adds a parameter setting.
     **/
    public void add(String name, String value)
    {

        Native.paramsSetSymbol(getContext().nCtx(), getNativeObject(),
                getContext().mkSymbol(name).getNativeObject(),
                getContext().mkSymbol(value).getNativeObject());
    }
    
    /**
     * A string representation of the parameter set.
     **/
    @Override
    public String toString()
    {
        return Native.paramsToString(getContext().nCtx(), getNativeObject());
    }

    Params(Context ctx)
    {
        super(ctx, Native.mkParams(ctx.nCtx()));
    }


    @Override
    void incRef() {
        Native.paramsIncRef(getContext().nCtx(), getNativeObject());
    }

    @Override
    void addToReferenceQueue() {
        getContext().getParamsDRQ().storeReference(getContext(), this);
    }
}

File: src/api/dotnet/Params.cs
/*++
Copyright (c) 2012 Microsoft Corporation

Module Name:

    Parameter.cs

Abstract:

    Z3 Managed API: Parameters

Author:

    Christoph Wintersteiger (cwinter) 2012-03-20

Notes:
    
--*/

using System.Diagnostics;
using System;

namespace Microsoft.Z3
{
    /// <summary>
    /// A Params objects represents a configuration in the form of Symbol/value pairs.
    /// </summary>
    public class Params : Z3Object
    {
        /// <summary>
        /// Adds a parameter setting.
        /// </summary>
        public Params Add(Symbol name, bool value)
        {
            Debug.Assert(name != null);

            Native.Z3_params_set_bool(Context.nCtx, NativeObject, name.NativeObject, (byte)(value ? 1 : 0));
            return this;
        }

        /// <summary>
        /// Adds a parameter setting.
        /// </summary>
        public Params Add(Symbol name, uint value)
        {
            Debug.Assert(name != null);

            Native.Z3_params_set_uint(Context.nCtx, NativeObject, name.NativeObject, value);
            return this;
        }

        /// <summary>
        /// Adds a parameter setting.
        /// </summary>
        public Params Add(Symbol name, double value)
        {
            Debug.Assert(name != null);            

            Native.Z3_params_set_double(Context.nCtx, NativeObject, name.NativeObject, value);
            return this;
        }

        /// <summary>
        /// Adds a parameter setting.
        /// </summary>
        public Params Add(Symbol name, string value)
        {
            Debug.Assert(value != null);

            Native.Z3_params_set_symbol(Context.nCtx, NativeObject, name.NativeObject, Context.MkSymbol(value).NativeObject);
            return this;
        }

        /// <summary>
        /// Adds a parameter setting.
        /// </summary>
        public Params Add(Symbol name, Symbol value)
        {
            Debug.Assert(name != null);
            Debug.Assert(value != null);

            Native.Z3_params_set_symbol(Context.nCtx, NativeObject, name.NativeObject, value.NativeObject);
            return this;
        }


        /// <summary>
        /// Adds a parameter setting.
        /// </summary>
        public Params Add(string name, bool value)
        {
            Native.Z3_params_set_bool(Context.nCtx, NativeObject, Context.MkSymbol(name).NativeObject, (byte)(value ? 1 : 0));
	    return this;
        }

        /// <summary>
        /// Adds a parameter setting.
        /// </summary>
        public Params Add(string name, uint value)
        {
            Native.Z3_params_set_uint(Context.nCtx, NativeObject, Context.MkSymbol(name).NativeObject, value);
	    return this;
        }

        /// <summary>
        /// Adds a parameter setting.
        /// </summary>
        public Params Add(string name, double value)
        {
            Native.Z3_params_set_double(Context.nCtx, NativeObject, Context.MkSymbol(name).NativeObject, value);
            return this;
        }

        /// <summary>
        /// Adds a parameter setting.
        /// </summary>
        public Params Add(string name, Symbol value)
        {
            Debug.Assert(value != null);

            Native.Z3_params_set_symbol(Context.nCtx, NativeObject, Context.MkSymbol(name).NativeObject, value.NativeObject);
            return this;
        }

        /// <summary>
        /// Adds a parameter setting.
        /// </summary>
        public Params Add(string name, string value)
        {
            Debug.Assert(name != null);
            Debug.Assert(value != null);

            Native.Z3_params_set_symbol(Context.nCtx, NativeObject, Context.MkSymbol(name).NativeObject, Context.MkSymbol(value).NativeObject);
            return this;
        }

        /// <summary>
        /// A string representation of the parameter set.
        /// </summary>
        public override string ToString()
        {
            return Native.Z3_params_to_string(Context.nCtx, NativeObject);
        }

        #region Internal
        internal Params(Context ctx)
            : base(ctx, Native.Z3_mk_params(ctx.nCtx))
        {
            Debug.Assert(ctx != null);
        }      

        internal override void IncRef(IntPtr o)
        {
            Native.Z3_params_inc_ref(Context.nCtx, o);
        }

        internal override void DecRef(IntPtr o)
        {
            lock (Context)
            {
                if (Context.nCtx != IntPtr.Zero)
                    Native.Z3_params_dec_ref(Context.nCtx, o);
            }
        }
        #endregion
    }
}
